File: src/main/java/in/costea/wiles/statements/ForStatement.java
package in.costea.wiles.statements;

import in.costea.wiles.builders.Context;
import in.costea.wiles.data.CompilationExceptionsCollection;
import in.costea.wiles.enums.SyntaxType;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class ForStatement extends AbstractStatement{
    public ForStatement(@NotNull Context oldContext) {
        super(oldContext.setWithinLoop());
    }

    @NotNull
    @Override
    public SyntaxType getType() {
        return SyntaxType.FOR;
    }

    @NotNull
    @Override
    public List<AbstractStatement> getComponents() {
        return List.of();
    }

    @NotNull
    @Override
    public CompilationExceptionsCollection process() {
        return new CompilationExceptionsCollection();
    }
}

File: wiles-base/src/main/java/wiles/parser/statements/ForStatement.kt
package wiles.parser.statements

import wiles.parser.builders.ExpectParamsBuilder.Companion.tokenOf
import wiles.parser.builders.ParserContext
import wiles.parser.statements.expressions.DefaultExpression
import wiles.shared.WilesException
import wiles.shared.AbstractStatement
import wiles.shared.WilesExceptionsCollection
import wiles.shared.SyntaxType
import wiles.shared.constants.ErrorMessages.IDENTIFIER_EXPECTED_ERROR
import wiles.shared.constants.Predicates
import wiles.shared.constants.Tokens.IN_ID

class ForStatement(oldContext: ParserContext) : AbstractStatement(oldContext) {
    private var identifierStatement : TokenStatement? = null
    private lateinit var inToken : TokenStatement
    private var inExpression : DefaultExpression = DefaultExpression(context)
    private val codeBlock = CodeBlockStatement(context)

    override val syntaxType: SyntaxType
        get() = SyntaxType.FOR

    override fun getComponents(): MutableList<AbstractStatement> {
        val list = mutableListOf<AbstractStatement>(identifierStatement?:return mutableListOf())
        list.add(inToken)
        list.add(inExpression)
        list.add(codeBlock)
        return list
    }

    override fun process(): WilesExceptionsCollection {
        val exceptions = WilesExceptionsCollection()
        try
        {
            identifierStatement = TokenStatement(transmitter.expect(tokenOf(Predicates.IS_IDENTIFIER)
                .withErrorMessage(IDENTIFIER_EXPECTED_ERROR)),context)

            val maybeToken = transmitter.expect(tokenOf(IN_ID))
            inToken = TokenStatement(maybeToken,context)
            inExpression = DefaultExpression(context)
            exceptions.addAll(inExpression.process())

            exceptions.addAll(codeBlock.process())
        }
        catch (ex : WilesException){
            exceptions.add(ex)
        }
        return exceptions
    }
}